crates/egui/src/containers/popup.rs

```diff
@@ -1,457 +1,517 @@
+        if let OpenKind::Memory { set, .. } = open_kind {
+            ctx.memory_mut(|mem| match set {
+                SetOpenState::DoNothing => {}
+                SetOpenState::Bool(open) => {
+                    if open {
+                        match self.anchor {
+                            PopupAnchor::PointerFixed => {
+                                mem.open_popup_at(id, hover_pos);
+                            }
+                            _ => mem.open_popup(id),
+                        }
+                    } else {
+                        mem.close_popup();
+                    }
+                }
+                SetOpenState::Toggle => {
+                    mem.toggle_popup(id);
+                }
+            });
+        }
+
+        if !open_kind.is_open(id, ctx) {
+            return None;
+        }
+
+        let (ui_kind, order) = match kind {
+            PopupKind::Popup => (UiKind::Popup, Order::Foreground),
+            PopupKind::Tooltip => (UiKind::Tooltip, Order::Tooltip),
+            PopupKind::Menu => (UiKind::Menu, Order::Foreground),
+        };
+
+        if kind == PopupKind::Popup {
+            ctx.pass_state_mut(|fs| {
+                fs.layers
+                    .entry(layer_id)
+                    .or_default()
+                    .open_popups
+                    .insert(id)
+            });
+        }
+
+        let anchor_rect = anchor.rect(id, ctx)?;
+
+        let (pivot, anchor) = best_align.pivot_pos(&anchor_rect, gap);
+
+        let mut area = Area::new(id)
+            .order(order)
+            .kind(ui_kind)
+            .pivot(pivot)
+            .fixed_pos(anchor)
+            .sense(sense)
+            .layout(layout);
+
+        if let Some(width) = width {
+            area = area.default_width(width);
+        }
+
+        let frame = frame.unwrap_or_else(|| Frame::popup(&ctx.style()));
+
+        let response = area.show(ctx, |ui| frame.show(ui, content).inner);
+
+        let should_close = |close_behavior| {
+            let should_close = match close_behavior {
+                PopupCloseBehavior::CloseOnClick => widget_clicked_elsewhere,
+                PopupCloseBehavior::CloseOnClickOutside => {
+                    widget_clicked_elsewhere && response.response.clicked_elsewhere()
+                }
+                PopupCloseBehavior::IgnoreClicks => false,
+            };
+
+            should_close || ctx.input(|i| i.key_pressed(Key::Escape))
+        };
+
+        match open_kind {
+            OpenKind::Open | OpenKind::Closed => {}
+            OpenKind::Bool(open, close_behavior) => {
+                if should_close(close_behavior) {
+                    *open = false;
+                }
+            }
+            OpenKind::Memory { close_behavior, .. } => {
+                if should_close(close_behavior) {
+                    ctx.memory_mut(|mem| mem.close_popup());
+                }
+            }
         }
```

Comment:
It would be nice with something like

```rust
if should_close(close_behavior) {
    open_kind.set_open(false);
}
```

…thought maybe for a later PR